temporal-sdk/src/main/java/io/temporal/internal/worker/TrackingSlotSupplier.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import com.uber.m3.tally.Scope;
+import io.temporal.worker.MetricsType;
+import io.temporal.worker.tuning.*;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Wraps a slot supplier and supplements it with additional tracking information that is useful to
+ * provide to all implementations. This type is used internally rather than {@link SlotSupplier}
+ * directly.
+ *
+ * @param <SI> The slot info type
+ */
+public class TrackingSlotSupplier<SI extends SlotInfo> {
+  private final SlotSupplier<SI> inner;
+  private final AtomicInteger issuedSlots = new AtomicInteger();
+  private final Map<SlotPermit, SI> usedSlots = new ConcurrentHashMap<>();
+  private final Scope metricsScope;
+
+  public TrackingSlotSupplier(SlotSupplier<SI> inner, Scope metricsScope) {
+    this.inner = inner;
+    this.metricsScope = metricsScope;
+    publishSlotsMetric();
+  }
+
+  public SlotPermit reserveSlot(SlotReservationData dat) throws InterruptedException {
+    SlotPermit p = inner.reserveSlot(createCtx(dat));
+    issuedSlots.incrementAndGet();
+    return p;
+  }
+
+  public Optional<SlotPermit> tryReserveSlot(SlotReservationData dat) {
+    Optional<SlotPermit> p = inner.tryReserveSlot(createCtx(dat));
+    if (p.isPresent()) {
+      issuedSlots.incrementAndGet();
+    }
+    return p;
+  }
+
+  public void markSlotUsed(SI slotInfo, SlotPermit permit) {
+    if (permit == null) {
+      throw new IllegalArgumentException(
+          "Permit cannot be null when marking slot as used. This is an SDK bug.");
+    }
+    if (usedSlots.put(permit, slotInfo) != null) {
+      throw new IllegalStateException("Slot is being marked used twice. This is an SDK bug.");
+    }
+    inner.markSlotUsed(new SlotMarkUsedContextImpl(slotInfo, permit));
+    publishSlotsMetric();
+  }
+
+  public void releaseSlot(SlotReleaseReason reason, SlotPermit permit) {
```

Comment:
What happens on a double release? I think right now it will silently pass?

Reply:
It will. There's not a great way to avoid that that I see, at least not without making some kind of publicly accessible `isReleased` field on the permit, since that's in worker rather than internal. Could be done though. Alternatively it can just be tested for with `CountingSlotSupplier` in the tests.

Reply:
Yeah I was thinking it was easier to check then maybe it actually is since it probably comes down to the route slot supplier. Maybe just testing `CountingSlotSupplier` is enough